gslib/context_config.py

```diff
@@ -0,0 +1,198 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Manages device context mTLS certificates."""
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import division
+from __future__ import unicode_literals
+
+import atexit
+import os
+import subprocess
+
+from boto import config
+
+import gslib
+
+# Maintain a single context configuration.
+_singleton_config = None
+
+
+class CertProvisionError(Exception):
+  """Represents errors when provisioning a client certificate."""
+  pass
+
+
+def _IsPemSectionMarker(line):
+  """Returns (begin:bool, end:bool, name:str)."""
+  if line.startswith('-----BEGIN ') and line.endswith('-----'):
+    return True, False, line[11:-5]
+  elif line.startswith('-----END ') and line.endswith('-----'):
+    return False, True, line[9:-5]
+  else:
+    return False, False, ''
+
+
+def _SplitPemIntoSections(contents, logger):
```

Comment:
I came across this helper implemented in google auth, can we just re-use that? https://github.com/googleapis/google-auth-library-python/blob/6407258956ec42e3b722418cb7f366e5ae9272ec/google/auth/transport/_mtls_helper.py#L90

Also, looks like they are using regex matching, which might be relatively easier to follow.

Reply:
Great find. I'll make some changes since the gcloud implementation accepts duplicate cert values and just goes with the latest, whereas the google-auth-library errors.

Reply:
Discussed offline. Splitting into lines and parsing like the gcloud implementation helps for catching nested matches.